compiler/src/iree/compiler/GlobalOptimization/PropagateDataLayout.cpp

```diff
@@ -0,0 +1,710 @@
+        status = ChangeStatus::CHANGED;
+      } else {
+        delete emptyTfLayout;
+      }
+      // Try to infer the transformation to the current value using the
+      // transformation from the neighboring value.
+      if (!neighborLayout->hasValidTransform()) {
+        continue;
+      }
+      auto *newLayout = new DataLayoutTransformation(*neighborLayout);
+      if (newLayout->transformLayout(neighbor, value)) {
+        // If there is already a known transformation to the current node, then
+        // try to combine the information from the new inferred layout with the
+        // known transformation, since not all transformations will carry the
+        // maximal information for a given value.
+        auto currentTf = newState.getDataLayoutTransformation(id);
+        if (currentTf && currentTf->hasValidTransform()) {
+          if (currentTf->combineLayout(*newLayout)) {
+            status = ChangeStatus::CHANGED;
+          }
+          delete newLayout;
+          continue;
+        }
+        // Otherwise, take the inferred transformation.
+        newState.setDataLayoutTransformation(id, newLayout);
+        status = ChangeStatus::CHANGED;
+      } else {
+        delete newLayout;
+      }
+    }
+  }
+
+  // If the newState did not pick up any layoutIDs from its neighbors, then
+  // this is the first node to be initialized with layoutIDs, and we need to
+  // walk the graph to find all reachable layouts from this node. These layouts
+  // will propagate through the neighboring nodes when the neighbors do a state
+  // update, so this should only happen once per connected graph.
+  if (newState.getLayoutIDs().size() == 0) {
+    newState.initializeTerminalNodeIDs(value);
+    walkAllConnectedTensorValues(
+        [&newState](Value v) { newState.initializeTerminalNodeIDs(v); });
+    if (newState.getLayoutIDs().size() > 0) {
+      status = ChangeStatus::CHANGED;
+    }
+  }
+  // Terminal nodes (sources of the layouts) initialize their corresponding
+  // layoutID with the identity transformation.
+  if (newState.initializeTerminalNodeLayouts(value)) {
+    status = ChangeStatus::CHANGED;
+  }
+
+  // Intermediate nodes add ValueElements for all neighboring tensor values.
+  if (newState.getNodeType() == DataLayoutNodeType::INTERMEDIATE) {
+    for (Value val : valueNeighbors) {
+      if (val != value) {
+        if (val.getDefiningOp<tensor::EmptyOp>()) {
+          continue;
+        }
+        auto &newNode = solver.getElementFor<GlobalDataLayoutValueElement>(
+            *this, Position::forValue(val), DFX::Resolution::REQUIRED);
+        solver.recordDependence(*this, newNode, DFX::Resolution::REQUIRED);
+      }
+    }
+  }
+
+  return status;
+}
+
+//===----------------------------------------------------------------------===//
+// Propagation pattern definitions
+//===----------------------------------------------------------------------===//
+
+class FoldCancellingUnPackPackOps final
```

Comment:
We should just add this upstream as a folding/canonicalization to pack/unpack operations.

Reply:
It's just calling `tensor::UnPackOp::canonicalize`, why not just populate with `tensor::UnPackOp::getCanonicalizationPatterns`.